### DISCUSSION TITLE
Form values reset to default after submit clicked

### DISCUSSION THREAD

**Discussion by gajo357** [react-hook-form/react-hook-form#2019] [Q&A] [answered]
**To Reproduce**
Using Ionic and latest versions of everything.
1. Change to input values to something illegal
2. Click the submit button
3. All input fields get reset to default

**Codesandbox link (Required)**
Using the [sandbox](https://codesandbox.io/s/29zur), just update the dependencies to latest and enable the Submit button even if the input is invalid.

`"dependencies": {
    "@ionic/react": "5.0.7",
    "@material-ui/core": "4.6.1",
    "react": "16.13.0",
    "react-dom": "16.13.0",
    "react-hook-form": "^6.0.0",
    "react-hook-form-input": "1.1.10",
    "react-scripts": "3.4.1",
    "react-select": "3.0.8"
  },`

`<IonButton type="submit">submit</IonButton>`

**Expected behavior**
I expected to continue where I have left off, not to be reset to initial values.

**Comment by bluebill1049** [comment]
you can't just take a codesandbox (from someone else) and update to the latest version (major version) and expect it's working. especially when there are breaking changes. the codesandbox was made probably before v5, `react-hook-form-input` is already part of the main repo. can you build one codesandbox by yourself with the latest version and see if you can reproduce any problem.

**Comment by gajo357** [comment]
Thanks @bluebill1049 for prompt reply.
I have created a [sandbox](https://codesandbox.io/s/react-hook-form-ionic-bug-7v3ur) from scratch and using `react-hook-form-input` in it.

**Comment by bluebill1049** [comment]
Don't use `react-hook-form-input`, please use `Controller` instead. Take a look this section https://react-hook-form.com/api/#Controller

there is a codesandbox as well: https://codesandbox.io/s/react-hook-form-v6-controller-qsd8r

**Reply by gajo357** [reply]
`Controller` does not have the `onChangeName` property and I do not know how to connect it to an Ionic input component (which triggers `onIonChange` event). Without it, the form's values never update.

**Comment by gajo357** [comment]
I have used the `Controller` initially, but then switched to `RHFInput` because the controller had same issues. 
It does not have `onChangeName` property, and the form does not re-render when I change the input.

It would be lovely to have a working sample with Ionic and `react-hook-form`, with or without `react-hook-form-input`.

**Comment by pogo19** [comment]
I've updated to v6 and apart from FormContext to FormProvider change it looked OK. But my reset in useEffect caused submit to send default values (in my case empty object) to onSubmit.

**Comment by bluebill1049** [comment]
> I've updated to v6 and apart from FormContext to FormProvider change it looked OK. But my reset in useEffect caused submit to send default values (in my case empty object) to onSubmit.

do you have code example? with `reset` problem?

**Comment by gajo357** [comment]
> I've updated to v6 and apart from FormContext to FormProvider change it looked OK. But my reset in useEffect caused submit to send default values (in my case empty object) to onSubmit.

I am not sure I understand your problem @pogo19 . In my [codesandbox](https://codesandbox.io/s/react-hook-form-ionic-bug-7v3ur) there is no `FormContext`, `FormProvider` nor `reset`.

**Comment by pogo19** [comment]
> > I've updated to v6 and apart from FormContext to FormProvider change it looked OK. But my reset in useEffect caused submit to send default values (in my case empty object) to onSubmit.
> 
> do you have code example? with `reset` problem?

Here it is: https://codepen.io/pogo19/pen/OJMzBMd?editors=0011
If you comment out the reset line, it works well. It used to work with 5.x.

Note that the reset only happens with initial form rendering, not during submit ;)

From another usage elsewhere in my code, I need to use that reset because I have the `defaultValues` stored in react state (first null, later updated from REST API) and I want to update the form with values on state change.

**Comment by pogo19** [comment]
> > I've updated to v6 and apart from FormContext to FormProvider change it looked OK. But my reset in useEffect caused submit to send default values (in my case empty object) to onSubmit.
> 
> I am not sure I understand your problem @pogo19 . In my [codesandbox](https://codesandbox.io/s/react-hook-form-ionic-bug-7v3ur) there is no `FormContext`, `FormProvider` nor `reset`.

I had a problem and it seemed to me it is similar to what I do experience after 5.x to 6.0.0 update of react-hook-form. If it is not related, I will open a new issue.

**Comment by bluebill1049** [comment]
I don't think there is an issue here, I will move this into the discussion.

**Comment by bluebill1049** [comment, marked as answer]
@pogo19 can you create a codesandbox for your problem? (keep it simple if possible)

**Reply by pogo19** [reply]
@bluebill1049 We had a long weekend over here in Czech Republic. Here it is: https://codesandbox.io/s/react-hook-form-reset-problem-37uir?file=/src/App.js

I hope I have simplified it to the bone. The problem persists. It doesn't work (values object printed to console is empty). If you comment out the `reset(..)` line, it works.

@gajo357 I've got rid of `FormProvider` and `useFormContext` in my example ;)

**Reply by gajo357** [reply]
Thanks for the effort. It gave me ideas and now everything works fine. Really nice library to use.

**Reply by bluebill1049** [reply]
<img width="949" alt="Screen Shot 2020-07-08 at 8 59 57 am" src="https://user-images.githubusercontent.com/10513364/86853205-68735500-c0f9-11ea-90fb-c989abf88f62.png">

Thanks, this was reported yesterday as an issue as well, I have updated the doc related to this problem.

**Reply by pogo19** [reply]
OK, thanks.

**Comment by qiluo** [comment]
To original question, I came into same issue. Then I figured out Controller v6 dropped onChangeName and replaced with render props, used render to connect inner input change and everything went ok
